extra/migrate_settings.rb

```diff
@@ -1,108 +1,230 @@
-#!/usr/bin/ruby
-#
-# Script to migrate forema-proxy settings.yml
-# to plugin form
-#
-# Greg Sutcliffe <gsutclif@redhat.com> 2014
-
+require 'pathname'
+require 'fileutils'
+require 'optparse'
 require 'yaml'
 
-def modules
-  [ :settings, :tftp, :dns, :dhcp, :puppet, :puppetca, :bmc, :chef, :realm ]
-end
+module ::Proxy
+  class Migrations
+    attr_reader :migrations, :past_migrations, :migration_state_file_path
+    attr_accessor :executed_migrations
 
-def known_options
-  {
-    :daemon             => :settings,
-    :daemon_pid         => :settings,
-    :log_file           => :settings,
-    :log_level          => :settings,
-    :port               => :settings,
-    :ssl_ca_file        => :settings,
-    :ssl_certificate    => :settings,
-    :ssl_private_key    => :settings,
-    :trusted_hosts      => :settings,
-    :virsh_network      => :settings,
-    :foreman_url        => :settings,
-    :settings_directory => :settings,
-    :http_port          => :settings,
-    :https_port         => :settings,
-    :use_cache          => :settings,
-    :cache_location     => :settings,
-    :puppetca_use_sudo  => :puppetca,
-    :puppetdir          => :puppetca,
-    :ssldir             => :puppetca,
-    :sudo_command       => :puppetca,
-    :customrun_args     => :puppet,
-    :customrun_cmd      => :puppet,
-    :freeipa_remove_dns => :realm
-  }
-end
+    def initialize(migration_state_file_path, past_migrations = [])
+      @past_migrations = past_migrations
+      @migration_state_file_path = migration_state_file_path
+      @executed_migrations = []
+    end
+
+    def load_past_migrations!
+      raise "Migration state file #{@migration_state_file_path} can't be found" unless File.exist?(@migration_state_file_path)
+      @past_migrations = File.read(@migration_state_file_path).split("\n")
+      self
+    end
+
+    def register_migration(migration_class)
+      @migrations ||= []
+      @migrations << migration_class
+    end
+
+    def new_migrations
+      # don't want to deal with namespaced constant resolution/loading
+      migrations.select {|m| !past_migrations.include?(m.name) }
+    end
+
+    def persist_migrations_state(result_dir_path)
+      migration_state = File.open(File.join(result_dir_path, "migration_state"), "w")
+      migration_state.write((past_migrations + executed_migrations.map {|m| m.class.name}).uniq.join("\n"))
+    ensure
+      migration_state.close unless migration_state.nil?
+    end
+
+    def modules_configuration_exist?
+      past_migrations.include?("MigrateMonolithicConfig")
```

Comment:
There's a bug when you're migrating the existing (say 1.8) config which *does* have a settings.d/ directory but this migration hasn't run (since our state file is empty).

This results in us trying to run MigrateMonolithicConfig, but copying the original config files into the source dir for the migration skips settings.d, because it runs this exist? method, which returns false.  We do actually need to copy the settings.d/ directory, if it exists, into the source dir else we lose it.

<pre>
✔ ~/code/foreman/smart-proxy [:74ed262|✚ 1…5] 
16:06 $ ll config/migration_state
-rw-rw-r--. 1 dcleal dcleal 0 May 22 11:00 config/migration_state
✔ ~/code/foreman/smart-proxy [:74ed262|✚ 1…5] 
16:06 $ ruby extra/migrate_settings.rb 
using:
  config file: /home/dcleal/code/foreman/smart-proxy/config/settings.yml,
  modules config dir: /home/dcleal/code/foreman/smart-proxy/config/settings.d,
  working dir: /home/dcleal/code/foreman/smart-proxy/tmp,
  migrations dir: /home/dcleal/code/foreman/smart-proxy/extra/migrations,
  migrations state path: /home/dcleal/code/foreman/smart-proxy/config/migration_state
      
Running migrations...
migrate_monolithic_config
copy_original_configuration("/home/dcleal/code/foreman/smart-proxy/tmp/migrate_monolithic_config/src", false)
✔ ~/code/foreman/smart-proxy [:74ed262|✚ 1…5] 
16:07 $ ll tmp/result/
total 12
-rw-rw-r--. 1 dcleal dcleal   23 May 27 16:07 migration_state
drwxrwxr-x. 2 dcleal dcleal 4096 May 27 15:59 settings.d
-rw-rw-r--. 1 dcleal dcleal  481 May 27 16:07 settings.yml
✔ ~/code/foreman/smart-proxy [:74ed262|✚ 1…5] 
16:07 $ ll tmp/result/settings.d/
total 0
✔ ~/code/foreman/smart-proxy [:74ed262|✚ 1…5] 
16:07 $ ll tmp/migrate_monolithic_config/*
tmp/migrate_monolithic_config/dst:
total 8
drwxrwxr-x. 2 dcleal dcleal 4096 May 27 15:59 settings.d
-rw-rw-r--. 1 dcleal dcleal  481 May 27 16:07 settings.yml

tmp/migrate_monolithic_config/src:
total 4
-rw-r--r--. 1 dcleal dcleal 442 May 27 16:07 settings.yml
✔ ~/code/foreman/smart-proxy [:74ed262|✚ 1…5] 
16:07 $ ll tmp/migrate_monolithic_config/dst/settings.d/
total 0
✔ ~/code/foreman/smart-proxy [:74ed262|✚ 1…5] 
</pre>